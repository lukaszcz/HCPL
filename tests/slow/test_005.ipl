
// laziness & higher-order functions

let filter = \lst \p { match lst with (%x, %t) -> if p x then (cons& x (filter t p)) else filter t p }
let take = \lst \n { if n == 0 then [] else match lst with (%h, %t) -> (cons h (take t (n - 1))) }
let nth = \lst \n { match lst with (%x, %t) -> if n == 1 then x else nth t (n - 1) }
let numbers = \n . cons& n (numbers (n + 1))

let divisible = \n \x . x mod n != 0
let eratostenes = \lst { let n = hd lst in (cons& n (eratostenes (filter (tl lst) (divisible n)))) }
let primes = eratostenes (numbers 2); // memory: linear, time: quadratic

let greater = \n \x . x > n

let f = \n \lst {
   let aux = \m \lst {
      if m == n then lst else aux (m + 1) (filter lst (greater m));
   };
   aux 0 lst
};

print (hd (f 3000 (numbers 0))); // memory: linear, time: quadratic
print (hd (filter (numbers 0) (\x . x > 10000000))); // memory & time: linear

print (take primes 10);
print (nth primes 100);
print (nth primes 1000);
print (nth primes 1000);
print (nth primes 5793);
print (nth primes 5793);
print (nth primes 10000);
print (nth primes 10000);
