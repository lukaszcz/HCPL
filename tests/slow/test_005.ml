
type 'a strm_t = Cons of 'a * ('a strm_t lazy_t)

let rec filter strm p =
  match strm with
  | Cons(x, t) -> if p x then Cons(x, lazy (filter (Lazy.force t) p)) else filter (Lazy.force t) p

let rec take strm n = if n = 0 then [] else match strm with Cons(h, t) -> h :: (take (Lazy.force t) (n - 1))

let rec nth strm n = match strm with Cons(x, t) -> if n = 1 then x else nth (Lazy.force t) (n - 1)

let rec numbers n = Cons(n, lazy (numbers (n + 1)))

let divisible n x = x mod n <> 0

let hd (Cons(x, _)) = x
let tl (Cons(_, t)) = Lazy.force t

let rec eratostenes strm =
  let n = hd strm
  in
  Cons(n, lazy (eratostenes (filter (tl strm) (divisible n))))

let primes = eratostenes (numbers 2)

let greater n x = x > n

let f n lst =
  let rec aux m lst =
    if m = n then lst else aux (m + 1) (filter lst (greater m))
  in
  aux 0 lst;;

let list_to_string f lst =
  let rec prn lst =
    match lst with
    | [x] -> f x
    | h :: t -> (f h) ^ ", " ^ prn t
    | [] -> ""
  in
  "[" ^ prn lst ^ "]"

let print_list lst = print_endline (list_to_string (fun n -> string_of_int n) lst);;

print_endline (string_of_int (hd (f 3000 (numbers 0))));
print_endline (string_of_int (hd (filter (numbers 0) (fun x -> x > 10000000))));

print_list (take primes 10);
print_endline (string_of_int (nth primes 100));
print_endline (string_of_int (nth primes 1000));
print_endline (string_of_int (nth primes 1000));
print_endline (string_of_int (nth primes 5793));
print_endline (string_of_int (nth primes 5793));
print_endline (string_of_int (nth primes 10000));
print_endline (string_of_int (nth primes 10000))
