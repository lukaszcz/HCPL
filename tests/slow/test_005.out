3000
10000001
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
541
7919
7919
57131
57131
104729
104729
